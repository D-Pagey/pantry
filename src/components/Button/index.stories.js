import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, number, select, text } from '@storybook/addon-knobs';
import { Button } from '.';

const variants = {
    'Default button': '',
    'Selected (selected)': 'selected',
    'UnSelected (unselected)': 'unselected',
    'Submit (submit)': 'submit'
};

storiesOf('Button', module)
    .addDecorator(withKnobs)
    .add('default', () => {
        const width = `${number('width (px)', 150)}px`;

        return (
            <Button variant={select('type (style of button)', variants, '')} width={width}>
                {text('children', 'Click me')}
            </Button>
        );
    });
