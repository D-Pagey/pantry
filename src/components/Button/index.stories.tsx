import React from 'react';
import { Story } from '@storybook/react/types-6-0';
import { Button, ButtonProps } from '.';

export default { title: 'Button', component: Button };

const Template: Story<ButtonProps> = (args) => <Button {...args}>Click me</Button>;

export const Primary = Template.bind({});

Primary.args = {};
