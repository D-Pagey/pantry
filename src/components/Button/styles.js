import styled, { keyframes } from 'styled-components';
import { colours } from '../../tokens';

export const Button = styled.button.attrs(({ type }) => ({
    type: type || 'button'
}))`
    background-color: ${colours.darkGreen100};
    border: 0;
    border-radius: 5px;
    color: ${colours.white};
    cursor: pointer;
    padding: 1rem;
    margin: ${({ margin }) => margin};
    min-width: 68px;
    width: ${({ size }) => size === 'sm' && 'max-content'};
`;

export const SecondaryButton = styled(Button)`
    background-color: ${colours.white};
    border: 1px solid ${colours.darkGreen100};
    color: ${colours.darkGreen100};
    margin: ${({ margin }) => margin};
`;

export const DisabledButton = styled(Button)`
    background-color: ${colours.grey};
    color: ${colours.white};
    cursor: not-allowed;
    margin: ${({ margin }) => margin};
`;

export const DestructiveButton = styled(Button)`
    background-color: ${colours.red};
`;

const ellipsis = keyframes`
  to {
      width: 1.25rem;    
    }
`;

export const LoadingButton = styled(Button)`
    background-color: ${colours.lightGreen};
    cursor: wait;
    padding: 1rem 2rem 1rem 1rem;

    &:after {
        animation: ${ellipsis} steps(4, end) 1000ms infinite;
        content: '...';
        display: inline-block;
        position: absolute;
        overflow: hidden;
        vertical-align: bottom;
        width: 0px;
    }
`;
