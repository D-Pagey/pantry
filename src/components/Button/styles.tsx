import styled from 'styled-components';

interface ButtonProps extends React.HTMLAttributes<HTMLElement> {
    readonly margin?: string;
    readonly width?: string;
}

export const Button = styled.button.attrs({
    type: 'button'
})<ButtonProps>`
    background-color: #288efc;
    border: 0;
    border-radius: 0.25rem;
    color: white;
    cursor: pointer;
    font-size: 1rem;
    height: 45px;
    line-height: 1.2;
    margin: ${({ margin }) => margin};
    width: ${({ width }) => width || '200px'};
`;

export const Selected = styled(Button)`
    background-color: linear-gradient(199.65deg, #288efc 0%, #2863fc 100%);
    border-radius: 6px;
    color: white;
    flex-grow: 1;
`;

export const UnSelected = styled(Button)`
    background-color: white;
    border-radius: 6px;
    color: black;
    flex-grow: 1;
`;

export const Submit = styled(Button).attrs({
    type: 'submit'
})``;
