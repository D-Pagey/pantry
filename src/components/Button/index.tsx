import React, { FC, ReactNode } from 'react';
import * as S from './styles';

type ButtonProps = {
    children: ReactNode;
    disabled?: boolean;
    margin?: string;
    onClick?: Function;
    secondary?: boolean;
    size?: 'sm' | 'm' | 'l';
    type?: string;
    destructive?: boolean;
};

export const Button: FC<ButtonProps> = ({ children, disabled, secondary, destructive, ...props }) => {
    if (disabled) {
        return (
            <S.DisabledButton disabled={disabled} {...props}>
                {children}
            </S.DisabledButton>
        );
    }

    if (secondary) {
        return <S.SecondaryButton {...props}>{children}</S.SecondaryButton>;
    }

    if (destructive) {
        return <S.DestructiveButton {...props}>{children}</S.DestructiveButton>;
    }

    return <S.Button {...props}>{children}</S.Button>;
};
