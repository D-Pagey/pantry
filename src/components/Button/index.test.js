import React from 'react';
import Button from '.';

const props = {
    testId: 'testButton'
};

describe('Button component', () => {
    it('should render', () => {
        const { container } = render(<Button {...props}>Click Me</Button>);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should render specified width button', () => {
        const width = '300px';
        const { getByText } = render(
            <Button width={width} {...props}>
                Wide Button
            </Button>
        );
        expect(getByText('Wide Button')).toHaveStyleRule('width', width);
    });

    it.each`
        variant         | backgroundColor
        ${'selected'}   | ${'linear-gradient(199.65deg,#288efc 0%,#2863fc 100%)'}
        ${'unselected'} | ${'white'}
    `('should render $variant styles', ({ variant, backgroundColor }) => {
        const { getByTestId } = render(
            <Button {...props} variant={variant}>
                Click Me
            </Button>
        );

        expect(getByTestId(props.testId)).toHaveStyleRule('background-color', backgroundColor);
    });
});
