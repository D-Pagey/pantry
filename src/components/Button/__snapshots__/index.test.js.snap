// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render 1`] = `
.c0 {
  background-color: #288efc;
  border: 0;
  border-radius: 0.25rem;
  color: white;
  cursor: pointer;
  font-size: 1rem;
  height: 45px;
  line-height: 1.2;
  width: 200px;
}

<button
  class="c0"
  data-testid="testButton"
  type="button"
  width=""
>
  Click Me
</button>
`;
